File: Sample/Source/NodeFunctions/System.py
# -*- encoding: utf-8 -*-
r"""\brief Blueprint system nodes: scene transitions, save/load, and game flow control."""

from dataclasses import fields
import os
from typing import Any, Callable
from Engine import Filters, Color
from Global import Manager, playVideo, System as GlobalSystem
from Global.Weather import WeatherType

SoundFilter = Filters.SoundFilter()
MusicFilter = Filters.MusicFilter()


class _TransitionCondition:
    def __init__(self) -> None:
        self._started = False

    def __call__(self) -> bool:
        if GlobalSystem.isTransitionPending() or GlobalSystem.isInTransition():
            self._started = True
            return False
        return self._started


def _isTransitionBackgroundFrozen() -> bool:
    return (
        not GlobalSystem.isTransitionBackgroundFreezePending()
        and GlobalSystem.isTransitionBackgroundFrozen()
    )


@Meta(
    DisplayName='LOC("EDIT_SOUND_FILTER")',
    DisplayDesc='LOC("EDIT_SOUND_FILTER_DESC")',
    DropBox={"attr": [field.name for field in fields(Filters.SoundFilter)]},
)
@ExecSplit(default=(None,))
def EditSoundFilter(attr: str, value: Any) -> None:
    setattr(SoundFilter, attr, value)


@Meta(
    DisplayName='LOC("EDIT_MUSIC_FILTER")',
    DisplayDesc='LOC("EDIT_MUSIC_FILTER_DESC")',
    DropBox={"attr": [field.name for field in fields(Filters.MusicFilter)]},
)
@ExecSplit(default=(None,))
def EditMusicFilter(attr: str, value: Any) -> None:
    setattr(MusicFilter, attr, value)


@Meta(DisplayName='LOC("PLAY_SOUND")', DisplayDesc='LOC("PLAY_SOUND_DESC")')
@ExecSplit(default=(None,))
def PlaySound(soundFileName: str, applyFilter: bool) -> None:
    global SoundFilter
    if applyFilter:
        Manager.playSE(soundFileName, SoundFilter)
    else:
        Manager.playSE(soundFileName)


@Meta(DisplayName='LOC("PLAY_MUSIC")', DisplayDesc='LOC("PLAY_MUSIC_DESC")')
@ExecSplit(default=(None,))
def PlayMusic(musicFileName: str, applyFilter: bool) -> None:
    global MusicFilter
    if applyFilter:
        Manager.playMusic("bgm", musicFileName, MusicFilter)
    else:
        Manager.playMusic("bgm", musicFileName)


@Meta(DisplayName='LOC("PLAY_VIDEO")', DisplayDesc='LOC("PLAY_VIDEO_DESC")')
@ExecSplit(default=(None,))
def PlayVideo(videoFileName: str, mute: bool, skipable: bool) -> None:
    videoPath = os.path.join(os.getcwd(), "Assets", "Videos", videoFileName)
    playVideo(videoPath, mute, skipable)


@Meta(
    DisplayName='LOC("FREEZE_TRANSITION_BACKGROUND")',
    DisplayDesc='LOC("FREEZE_TRANSITION_BACKGROUND_DESC")',
)
@Latent(Frozen=(True,))
def FreezeTransitionBackground() -> Callable[[], bool]:
    r"""\brief Freeze the current frame and wait until it is ready for a transition.

    - \return A condition callable that becomes True when the frame has been captured.
    """
    GlobalSystem.freezeTransitionBackground()
    return _isTransitionBackgroundFrozen


@Meta(DisplayName='LOC("REQUEST_TRANSITION")', DisplayDesc='LOC("REQUEST_TRANSITION_DESC")')
@Latent(Finished=(True,))
def RequestTransition(transitionName: str, transitionTime: float) -> Callable[[], bool]:
    r"""\brief Request a screen transition and wait until it finishes.

    - \param transitionName Optional transition texture filename.
    - \param transitionTime Transition duration in seconds.
    - \return A condition callable that becomes True when the transition is finished.
    """
    GlobalSystem.requestTransition(transitionName, float(transitionTime))
    return _TransitionCondition()


@Meta(
    DisplayName='LOC("SET_BGM_FILTER")',
    DisplayDesc='LOC("SET_BGM_FILTER_DESC")',
    DropBox={"attr": [field.name for field in fields(Filters.MusicFilter)]},
)
@ExecSplit(default=(None,))
def SetBgmFilter(attr: str, value: Any) -> None:
    scene = GlobalSystem.getScene()
    if scene and hasattr(scene, "setBgmFilter"):
        scene.setBgmFilter(attr, value)


@Meta(
    DisplayName='LOC("SET_BGS_FILTER")',
    DisplayDesc='LOC("SET_BGS_FILTER_DESC")',
    DropBox={"attr": [field.name for field in fields(Filters.SoundFilter)]},
)
@ExecSplit(default=(None,))
def SetBgsFilter(attr: str, value: Any) -> None:
    scene = GlobalSystem.getScene()
    if scene and hasattr(scene, "setBgsFilter"):
        scene.setBgsFilter(attr, value)


@Meta(DisplayName='LOC("FLASH_SCREEN")', DisplayDesc='LOC("FLASH_SCREEN_DESC")')
@ExecSplit(default=(None,))
def FlashScreen(red: int, green: int, blue: int, alpha: int, duration: float) -> None:
    GlobalSystem.flashScreen(Color(int(red), int(green), int(blue), int(alpha)), float(duration))


@Meta(DisplayName='LOC("STOP_FLASH_SCREEN")', DisplayDesc='LOC("STOP_FLASH_SCREEN_DESC")')
@ExecSplit(default=(None,))
def StopFlashScreen() -> None:
    GlobalSystem.stopFlash()


@Meta(DisplayName='LOC("SCREEN_SHAKE")', DisplayDesc='LOC("SCREEN_SHAKE_DESC")')
@ExecSplit(default=(None,))
def ScreenShake(power: float, speed: float, duration: float) -> None:
    GlobalSystem.startShake(float(power), float(speed), float(duration))


@Meta(DisplayName='LOC("STOP_SCREEN_SHAKE")', DisplayDesc='LOC("STOP_SCREEN_SHAKE_DESC")')
@ExecSplit(default=(None,))
def StopScreenShake() -> None:
    GlobalSystem.stopShake()


@Meta(
    DisplayName='LOC("SET_WEATHER")',
    DisplayDesc='LOC("SET_WEATHER_DESC")',
    DropBox={"weatherType": WeatherType.dropBoxItems()},
)
@ExecSplit(default=(None,))
def SetWeather(weatherType: WeatherType, power: int, maxCount: int) -> None:
    GlobalSystem.setWeather(WeatherType.coerce(weatherType), int(power), int(maxCount))


@Meta(DisplayName='LOC("CLEAR_WEATHER")', DisplayDesc='LOC("CLEAR_WEATHER_DESC")')
@ExecSplit(default=(None,))
def ClearWeather() -> None:
    GlobalSystem.clearWeather()

File: Sample/Scripts/Source/NodeFunctions/System.lua
local Engine = require("Engine")
local GlobalCore = require("GlobalCore")
local GlobalFunctions = require("GlobalFunctions")
local LocaleCore = require("Source.Locale.Core")
local Context = require("Source.NodeFunctions.Context")

local AudioManager = GlobalCore.AudioManager
local AudioEffects = GlobalCore.AudioEffects
local ManagerFunctions = GlobalFunctions.Manager
local GlobalSystem = GlobalCore.System
---@type fun(value: string): string
local LOC = LocaleCore.ApplyStringLocaleFormat
local System = {}

---@param constructor fun(values: table): Engine.SoundFilter
---@return Engine.SoundFilter
local function constructSoundFilter(constructor)
    ---@cast constructor function
    return constructor()
end

---@param constructor fun(values: table): Engine.MusicFilter
---@return Engine.MusicFilter
local function constructMusicFilter(constructor)
    ---@cast constructor function
    return constructor()
end

System.SoundFilter = constructSoundFilter(Engine.SoundFilter.new)
System.MusicFilter = constructMusicFilter(Engine.MusicFilter.new)

local TransitionCondition = {}

function TransitionCondition:init()
    ---@type boolean
    self._started = false
end

function TransitionCondition:poll()
    if GlobalSystem.isTransitionPending() or GlobalSystem.isInTransition() then
        self._started = true
        return false
    end
    return self._started
end

function TransitionCondition:isFinished()
    return self._started and not GlobalSystem.isTransitionPending() and not GlobalSystem.isInTransition()
end

local FinalTransitionCondition = class(TransitionCondition)
FinalTransitionCondition.__call = TransitionCondition.poll

local FrozenCondition = {}

function FrozenCondition.poll()
    return not GlobalSystem.isTransitionBackgroundFreezePending() and GlobalSystem.isTransitionBackgroundFrozen()
end

function FrozenCondition:isFinished()
    return self:poll()
end

local FinalFrozenCondition = class(FrozenCondition)
FinalFrozenCondition.__call = FrozenCondition.poll

function System.EditSoundFilter(attr, value)
    System.SoundFilter[attr] = value
end

function System.EditMusicFilter(attr, value)
    System.MusicFilter[attr] = value
end

function System.SetEffect(audioType, effect)
    local effectName = assert(AudioEffects[effect], "Unknown audio effect: " .. effect)
    AudioManager.setEffect(audioType, effectName)
end

function System.PlaySound(soundFileName, applyFilter)
    applyFilter = applyFilter == nil and false or applyFilter
    if applyFilter then
        ManagerFunctions.playSE(soundFileName, System.SoundFilter)
    else
        ManagerFunctions.playSE(soundFileName)
    end
end

function System.PlayMusic(musicFileName, applyFilter)
    applyFilter = applyFilter == nil and false or applyFilter
    local scene = GlobalSystem.getScene()
    local musicFilter = applyFilter and System.MusicFilter or nil
    if bool(scene) then
        ---@cast scene Source.Scenes.SceneMap.SceneMap
        scene:playBgm(musicFileName, musicFilter)
        return
    end
    local music = ManagerFunctions.playMusic("BGM", musicFileName, musicFilter)
    if music ~= nil then
        music:setLooping(true)
    end
end

function System.PlayVideo(videoFileName, mute, skipable)
    mute = mute == nil and false or mute
    skipable = skipable == nil and true or skipable
    local videoPath = os.path.join(os.getcwd(), "Assets", "Videos", videoFileName)
    _G.playVideo(videoPath, mute, skipable)
end

function System.FreezeTransitionBackground()
    GlobalSystem.freezeTransitionBackground()
    return FinalFrozenCondition.new()
end

function System.RequestTransition(transitionName, transitionTime)
    transitionName = transitionName == nil and "" or transitionName
    transitionTime = transitionTime == nil and 1.0 or transitionTime
    GlobalSystem.requestTransition(transitionName, tonumber(transitionTime))
    return FinalTransitionCondition.new()
end

function System.SetBgmFilter(attr, value)
    Context.requireSceneMap():setBgmFilter(attr, value)
end

function System.SetBgsFilter(attr, value)
    Context.requireSceneMap():setBgsFilter(attr, value)
end

function System.FlashScreen(red, green, blue, alpha, duration)
    red = red == nil and 255 or red
    green = green == nil and 255 or green
    blue = blue == nil and 255 or blue
    alpha = alpha == nil and 255 or alpha
    duration = duration == nil and 0.5 or duration
    GlobalSystem.flashScreen(
        sf.Color.new((math.modf(red)), (math.modf(green)), (math.modf(blue)), (math.modf(alpha))), tonumber(duration)
    )
end

function System.StopFlashScreen()
    GlobalSystem.stopFlash()
end

function System.ChangeScreenTone(red, green, blue, gray, duration)
    red = red == nil and 0 or red
    green = green == nil and 0 or green
    blue = blue == nil and 0 or blue
    gray = gray == nil and 0 or gray
    duration = duration == nil and 0.0 or duration
    GlobalSystem.changeScreenTone(tonumber(red), tonumber(green), tonumber(blue), tonumber(gray), tonumber(duration))
end

function System.ClearScreenTone(duration)
    duration = duration == nil and 0.0 or duration
    GlobalSystem.clearScreenTone(tonumber(duration))
end

function System.ScreenShake(power, speed, duration)
    power = power == nil and 4.0 or power
    speed = speed == nil and 10.0 or speed
    duration = duration == nil and 0.5 or duration
    GlobalSystem.startShake(tonumber(power), tonumber(speed), tonumber(duration))
end

function System.StopScreenShake()
    GlobalSystem.stopShake()
end

local function coerceWeatherType(weatherType)
    local names = { "NONE", "RAIN", "STORM", "SNOW" }
    for _, name in ipairs(names) do
        if weatherType == LOC("WEATHER_TYPE_" .. name) then
            return GlobalCore[name]
        end
    end
    return GlobalCore.coerce(weatherType)
end

function System.SetWeather(weatherType, power, maxCount)
    power = power == nil and 40 or power
    maxCount = maxCount == nil and 80 or maxCount
    GlobalSystem.setWeather(coerceWeatherType(weatherType), (math.modf(power)), (math.modf(maxCount)))
end

function System.ClearWeather()
    GlobalSystem.clearWeather()
end

return System
